[NEW BONUS GUIDE] Samourai Dojo (#1096)
raspibolt/raspibolt: guide/bonus/bitcoin/Dojo.md

Samourai Dojo requires several new passwords. They should be unique and very secure, at least 12 characters in length. Do not use uncommon special characters, spaces, or quotes (‘ or “).

```sh
[ F ] MYSQL_ROOT_PASSWORD
[ G ] MYSQL_PASSWORD
[ H ] NODE_API_KEY_1, NODE_API_KEY_2
[ I ] NODE_ADMIN_KEY
[ J ] NODE_JWT_SECRET
```

Store a copy of your passwords somewhere safe (preferably in an open-source password manager like [KeePassXC](https://keepassxc.org/){:target="_blank"}) or whatever password manager you're already using

### Node.js

To run Dojo, we need to run Node.js v16 or above.

* With user "admin", let's check our version of Node.js running on the node

```sh
$ node -v
> v16.13.1
```

* If Node.js is not installed, follow [this guide](../../bitcoin/blockchain-explorer.md#install-nodejs) to install it. If the version is v14 or older, update it following [this tutorial](https://phoenixnap.com/kb/update-node-js-version){:target="_blank"}.

### npm & PM2

npm is the default package manager for the JavaScript runtime environment and PM2 is a process manager for the JavaScript runtime Node.js

* Install npm package manager and update it to the latest version

```sh

> $ sudo apt install npm

Reviewer: After following the guide to install node.js this command gave me a error.
I think installing nodejs already installs npm, so you should check if npm is installed and update it

---
layout: default
title: Samourai Dojo
parent: + Bitcoin
grand_parent: Bonus Section
nav_exclude: true
has_children: false
has_toc: false
---

## Bonus guide: Samourai Dojo
{: .no_toc }

---

[Samourai Dojo](https://code.samourai.io/dojo/samourai-dojo){:target="_blank"} is the backing server for Samourai Wallet. Provides HD account & loose addresses (BIP47) balances & transactions lists. 
Provides unspent output lists to the wallet. PushTX endpoint broadcasts transactions through the backing bitcoind node.

Difficulty: Medium
{: .label .label-yellow }

Status: Not tested v3
{: .label .label-yellow }

![dojo_profile](../../../images/dojo_profile.png)

---

Table of contents
{: .text-delta }

1. TOC
{:toc}

---

## Requirements

* Minimum RAM: 4 GB
* Bitcoin Core
* Fulcrum
* Node.js v16+
* Nginx

---

## Preparations

### Write down your passwords 

Samourai Dojo requires several new passwords. They should be unique and very secure, at least 12 characters in length. Do not use uncommon special characters, spaces, or quotes (‘ or “).

```sh
[ F ] MYSQL_ROOT_PASSWORD
[ G ] MYSQL_PASSWORD
[ H ] NODE_API_KEY_1, NODE_API_KEY_2
[ I ] NODE_ADMIN_KEY
[ J ] NODE_JWT_SECRET
```

> Store a copy of your passwords somewhere safe (preferably in an open-source password manager like [KeePassXC](https://keepassxc.org/){:target="_blank"}) or whatever password manager you're already using

Reviewer 2: ```suggestion
Store a copy of your passwords somewhere safe (preferably in an open-source password manager like [KeePassXC](https://keepassxc.org/){:target="_blank"}) or whatever password manager you're already using)
```

missing end bracket

### Node.js

To run Dojo, we need to run Node.js v16 or above.

* With user "admin", let's check our version of Node.js running on the node

```sh
$ node -v
> v16.13.1
```

* If Node.js is not installed, follow [this guide](../../bitcoin/blockchain-explorer.md#install-nodejs) to install it. If the version is v14 or older, update it following [this tutorial](https://phoenixnap.com/kb/update-node-js-version){:target="_blank"}.

### npm & PM2

npm is the default package manager for the JavaScript runtime environment and PM2 is a process manager for the JavaScript runtime Node.js

* With user "admin", let's check our version of npm running on the node

```sh
$ npm -v
> 8.19.2
```

* If not already installed, install npm package manager and update it to latest version

```sh

> $ sudo apt install npm

Reviewer 2: ```suggestion
$ sudo apt update
$ sudo apt install npm
```

Probably worth updating the packake list before installing npm

$ sudo npm install latest-version
```

* Install PM2 process manager

```sh
$ sudo npm i -g pm2
```

---

## Installation

### Download Samourai Dojo

* As a user “admin” move to a temporary directory which is cleared on reboot

```sh
$ cd /tmp/
```

* Get the latest download links at [code.samourai.io/dojo](https://code.samourai.io/dojo/samourai-dojo/-/releases){:target="_blank"}. They change with each update

```sh

> $ sudo wget https://code.samourai.io/dojo/samourai-dojo/-/archive/v1.17.0/samourai-dojo-v1.17.0.tar.gz
> $ sudo tar -xvf samourai-dojo-v1.17.0.tar.gz

Reviewer 2: ```suggestion
  $ wget https://code.samourai.io/dojo/samourai-dojo/-/archive/v1.17.0/samourai-dojo-v1.17.0.tar.gz
  $ wget https://code.samourai.io/dojo/samourai-dojo/uploads/0c9f3d5d07396019f3bfb6449fe11e48/samourai-dojo-v1.17.0-fingerprints.txt
  $ wget https://code.samourai.io/dojo/samourai-dojo/uploads/4ad2b6e5c8a64f4a44bfd1ef12a5b1c8/samourai-dojo-v1.17.0-fingerprints.txt.sig

* Calculate the checksum of the binary you've downloaded and compare it to the one provided in the  fingerprints text file

  ```sh
  $ sha256sum --ignore-missing --check samourai-dojo-v1.17.0-fingerprints.txt
  > samourai-dojo-v1.17.0.tar.gz: OK
  ```

* Import the GPG public key of the developper that signed the fingerprints file

  ```sh
  $ curl https://keys.openpgp.org/vks/v1/by-fingerprint/377DE507FDC47F020099E342CFA54E4C0CD58DF0 | gpg --import
  ```

* Verify that the fingerprints file has actually been signed by that developper

  ```sh
  $ gpg --verify samourai-dojo-v1.17.0-fingerprints.txt.sig
  > gpg: assuming signed data in 'samourai-dojo-v1.17.0-fingerprints.txt'
  > gpg: Signature made Fri Aug  5 13:59:09 2022 BST
  > gpg:                using RSA key 377DE507FDC47F020099E342CFA54E4C0CD58DF0
  > gpg: Good signature from "pavel.sevcik@protonmail.com <pavel.sevcik@protonmail.com>" [unknown]
  > [...]
  ```

* If the signature checks out, unpack the binary

  ```sh
  $ tar -xvf samourai-dojo-v1.17.0.tar.gz
  ```

  ```

* Added some steps to verify the authenticity of the binary
* Removed sudo for wget and tar, not required

  [ J ] NODE JWT SECRET
  ```

### Node.js

To run Dojo, we need to run Node.js

* With user "admin", let's check our version of Node.js running on the node. If the version is v14 or older, update it following [this tutorial](https://phoenixnap.com/kb/update-node-js-version){:target="_blank"}.

  ```sh
  $ node -v
  > v16.13.1
  ```

* If Node.js is not installed, add the Node.js package repository from user “admin” and install Node.js using the apt package manager

  ```sh
  $ curl -fsSL https://deb.nodesource.com/setup_16.x | sudo -E bash -
  $ sudo apt install nodejs
  ```

### Node Package Manager & PM2 

Node Package Manager (npm) is the default package manager for the JavaScript runtime environment and PM2 is a process manager for the JavaScript runtime Node.js

* With user "admin", let's check our version of npm running on the node
 
  ```sh
  $ npm -v
  > 8.19.2
  ```

* If not already installed, install npm package manager

  ```sh
  $ sudo apt update
  $ sudo apt install npm
  ```

* Update npm to the latest version

  ```sh
  $ sudo npm install latest-version
  ```

* Install PM2 process manager

  ```sh
  $ sudo npm i -g pm2
  ```

### For Electrs

> If you are using Electrum server instead of Fulcrum, it is necessary to make following change inside Bitcoind configuration file. Fulcrum users can skip this step as it is already done.

Reviewer 2: ```suggestion
If you are using Electrs instead of Fulcrum, it is necessary to make following change inside Bitcoind configuration file. Fulcrum users can skip this step as it is already done.
```

Electrum server is too generic as Fulcrum is also one (afaik)

* Open `bitcoin.conf`

  ```sh
  $ sudo nano /home/bitcoin/.bitcoin/bitcoin.conf
  ```

* Add following line preferably under "# Connections" or at the end of the file 

  ```sh

>   zmqpubhashblock=tcp://0.0.0.0:8433  # For Fulcrum/Dojo

Reviewer 2: ```suggestion
  zmqpubhashblock=tcp://0.0.0.0:8433  # For Dojo
```

Removing Fulcrum in comment as this section is for Electrs users

  ```sh
  $ cd /tmp/
  ```

* Get the latest download links at [code.samourai.io/dojo](https://code.samourai.io/dojo/samourai-dojo/-/releases){:target="_blank"}. They change with each update

  ```sh
  $ wget https://code.samourai.io/dojo/samourai-dojo/-/archive/v1.17.0/samourai-dojo-v1.17.0.tar.gz
  $ wget https://code.samourai.io/dojo/samourai-dojo/uploads/0c9f3d5d07396019f3bfb6449fe11e48/samourai-dojo-v1.17.0-fingerprints.txt
  $ wget https://code.samourai.io/dojo/samourai-dojo/uploads/4ad2b6e5c8a64f4a44bfd1ef12a5b1c8/samourai-dojo-v1.17.0-fingerprints.txt.sig
  ```

* Calculate the checksum of the binary you've downloaded and compare it to the one provided in the fingerprints text file

  ```sh
  $ sha256sum --ignore-missing --check samourai-dojo-v1.17.0-fingerprints.txt
  > samourai-dojo-v1.17.0.tar.gz: OK
  ```

* Import the GPG public key of the developper that signed the fingerprints file

  ```sh
  $ curl https://keys.openpgp.org/vks/v1/by-fingerprint/377DE507FDC47F020099E342CFA54E4C0CD58DF0 | gpg --import
  ```

* Verify that the fingerprints file has actually been signed by that developper

  ```sh

>   $ gpg --verify samourai-dojo-v1.17.0-fingerprints.txt.sig > gpg: assuming signed data in 'samourai-dojo-v1.17.0-fingerprints.txt' > gpg: Signature made
>   Fri   Aug 5 13:59:09 2022 BST > gpg: using RSA key 377DE507FDC47F020099E342CFA54E4C0CD58DF0 > gpg: Good signature from "pavel.sevcik@protonmail.com
>   <pavel.sevcik@protonmail.com>" [unknown] > [...]

Reviewer 2: ```suggestion
  $ gpg --verify samourai-dojo-v1.17.0-fingerprints.txt.sig 
  > gpg: assuming signed data in 'samourai-dojo-v1.17.0-fingerprints.txt' 
  > gpg: Signature made Fri   Aug 5 13:59:09 2022 BST 
  > gpg: using RSA key 377DE507FDC47F020099E342CFA54E4C0CD58DF0 
  > gpg: Good signature from "pavel.sevcik@protonmail.com <pavel.sevcik@protonmail.com>" [unknown] 
  > [...]
```

fixing formatting of output lines

  ```sh
  [...]
  const INTERPRETER = 'node' // OR binary name like `node`
  ```

* Exit "dojo" user session

  ```sh
  $ exit
  ```

### Dependencies

* With user "admin", install necessary dependencies while inside the Dojo folder. 

  ```sh
  $ cd /opt/dojo/
  $ sudo npm install --only=prod
  > npm WARN config only Use `--omit=dev` to omit dev dependencies from the install.
    debug2: channel 0: window 999342 sent adjust 49234 Completed in 1131ms
    [...]
  ```

* Import Dojo scripts to the MariaDB database. Change [ F ] MYSQL_ROOT_PASSWORD to your password (must stay inside quotes: "password")

  ```sh
  $ sudo mysql -u"root" -p"[ F ] MYSQL_ROOT_PASSWORD" "dojo_db" < ./db-scripts/1_db.sql.tpl -v
  ```
  ```sh
    --------------
  > /*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */
    --------------
    [...]
  ```

### Tor Hidden Service

Tor is used to access "Dojo API and Maintanence tool" and to reach Dojo in an anonymous way.

* Edit "torrc" file

  ```sh
  $ sudo nano /etc/tor/torrc
  ```

* Paste following values at the end of your `torrc` file

  ```sh
  # Dojo hidden service
  SocksPort 9050
  SocksPolicy accept 127.0.0.1
  SocksPolicy reject *

> 

Author: If they are default, there is no reason for them:) did not know that

### Rescan public keys using Dojo Maintenance Tool (optional)
If no balance is shown in your samourai wallet, it is neccessary to rescan public keys as they are not tracked by Dojo yet.

* Log into `Dojo API and Maintenance Tool` using Tor browser
* Move to `xpubs tool`, under `Tools` section
* In samourai wallet go to `Settings > Wallet`, here you can find your public keys
* Copy and paste all "zpubs" into `xpub tool` and rescan each public key separately

Samourai wallet uses zpubs by default, however if you use other address format than "bc", it is neccessary to rescan other pubs as well

---

## For the future: Dojo upgrade 

* Switch to "dojo" user

  ```sh
  $ sudo su - dojo
  ```

* Stop Dojo

  ```sh
  $ pm2 stop mainnet
  ```

* Following the installation section, download and install latest Dojo version, start Dojo 

  ```sh
  $ pm2 start mainnet
  ```

---

## Uninstall

### Uninstall Dojo

* As user "admin", stop and remove Dojo service

  ```sh
  $ sudo systemctl stop pm2-dojo.service 
  $ sudo systemctl disable pm2-dojo.service 
  $ sudo rm /etc/systemd/system/pm2-dojo.service
  ```

* Remove Dojo directory and Dojo user

  ```sh

>   $ sudo rm -R /data/dojo

Reviewer 2: ```suggestion
  $ sudo rm -R /opt/dojo
```

  $ sudo userdel -r dojo
  ```

### Remove Tor Hidden Service

* Remove Tor configuration. Comment or delete following lines

  ```sh
  $ sudo nano /etc/tor/torrc
  ```

  ```sh
  # Dojo hidden service

>   #SocksPort 9050
>   #SocksPolicy accept 127.0.0.1
>   #SocksPolicy reject *
> 

Reviewer 2: ```suggestion
```

See comment above (depending if you removed them above)
